### PR TITLE
Backport package stage from refactoring_sandbox

### PR DESCRIPTION
**Pull request by kxl-adsk**

### CODE REVIEW

**Comment by kxl-adsk** [comment]
@murphyeoin, @mattyjams this is something that you are not using, but feel free to add review.

File: build.py [hunk 1]
```diff
@@ -274,6 +274,49 @@ def RunCTest(context, extraArgs=None):
...
+        shutil.make_archive(pkgName, 'zip', installDir)
+
+        # copy zip file to package directory
+        try:
+            if os.path.isdir(pkgDir):
+                testFile = os.path.join(pkgDir, "canwrite")
+                open(testFile, "w").close()
+                os.remove(testFile)
+            else:
+                os.makedirs(pkgDir)
```
Comment (mattyjams) [inline build.py]: Why not just let the `shutil.copy()` below raise a permissions error and truncate this to just create the directory if it doesn't exist?

File: build.py [hunk 2]
```diff
@@ -274,6 +274,49 @@ def RunCTest(context, extraArgs=None):
...
+
+            for file in os.listdir(buildDir):
+                if file.endswith(".zip"):
+                    zipFile = os.path.join(buildDir, file)
+                    shutil.copy(zipFile, pkgDir)
+        except Exception as e:
+            PrintError("Could not write to directory {dir}. Change permissions "
+                       "or choose a different location to install to."
+                       .format(dir=dir))
```
Comment (mattyjams) [inline build.py]: Maybe include the exception message in the error message for some additional context?

Comment (HamedSabri-adsk) [reply build.py]: Good point!  sure, I will have a change based on these feedbacks.

**Pull request merged**